File: longestSubString.cpp
// problem 3
#include<iostream>
#include<set>
using namespace std;

class Solution
{
    public:
        Solution(){}
        int lengthOfLongestSubstring(string s) 
		{	
			set<char> window = {};
			int left_index = 0;
			int longest_len = 0;

			for (int r_index = 0; r_index < s.length(); r_index++)
			{
				while (window.find(s[r_index]) != window.end())
				{
					window.erase(s[left_index]);
					left_index++;
				}
				window.insert(s[r_index]);
				longest_len = (longest_len > window.size()) ? longest_len : window.size();
			}
			return longest_len;
        }
};

int main(int argc, char const *argv[])
{
    Solution *solver = new Solution();
    int result = solver->lengthOfLongestSubstring("abcabcbb");
    cout << result << endl;
    return 0;
}

File: longestSubString.java
import java.util.HashSet;

// problem 3
public class longestSubString {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> window = new HashSet<>();
        int lIndex = 0;
        int longestLen = 0;
        
        for (int rIndex = 0; rIndex < s.length(); rIndex++) {
            // as long as duplicate is found in window
            while (window.contains(s.charAt(rIndex))) {
                window.remove(s.charAt(lIndex));
                lIndex++;
            } // this is needed to guarantee our window does not contain duplicates
            window.add(s.charAt(rIndex));
            longestLen = (longestLen > window.size()) ? longestLen : window.size();
        }

        return longestLen;
    }

    public static void main(String[] args) {
        longestSubString solution = new longestSubString();
        int answer = solution.lengthOfLongestSubstring("abcabcdbb");
        System.out.println(answer);
    }
}
